Update FAQ to reflect XMTP v3 as the current state (#26)
xmtp/docs-xmtp-org: docs/pages/get-started/faq.md

### How should I handle the XMTP onboarding flow in my app?

In your app onboarding flow, enable your users to activate XMTP messaging. User access to messaging can help with app engagement and re-engagement.

For example, here is a prompt to activate XMTP DMs in the onboarding flow to [claim a Lens handle](https://claim.lens.xyz/):

![activatedmsscreen-width-500px](https://raw.githubusercontent.com/xmtp/docs-xmtp-org/main/docs/pages/img/activate-dms-screen.png)

In your app onboarding flow, also request user permission to display app-specific push notifications. Enabling push notifications can help with user engagement and retention by allowing your app to reach users outside of active sessions. 

Push notifications are particularly valuable for messaging apps, as they help ensure users don't miss important communications and can respond promptly to new messages.

> ### Is there a way to get a list of all current XMTP-enabled wallets?

Reviewer: Don't have a strong opinion, but do we need to display this question so high up on the list?

Author: Thank you, @richardhuaaa - such a great callout. I've moved it to the end of the Build section. Just one more question down - but a good move nonetheless!

XMTP doesn't provide a direct method to fetch all XMTP-enabled wallet addresses in bulk. You might consider using a third-party service to do so instead of building the functionality yourself.

> [Blaze CRM](https://www.withblaze.app/) is one known service. It compiles lists of XMTP addresses by querying the XMTP `canMessage` API over time.

Author: Are there other services that offer this today?

The XMTP ecosystem is continually evolving. If you know of other reliable services providing this functionality, please consider sharing them in the [XMTP Community Forums](https://community.xmtp.org/c/development/third-party-tools/74).

## Network

### Is the XMTP network decentralized?

Currently, [Ephemera](https://ephemerahq.com/) (the company) operates all of the network nodes in the two available XMTP network environments: `dev` and `production`.

These network nodes operate in US jurisdiction in compliance with Office of Foreign Assets Control (OFAC) sanctions and Committee on Foreign Investment in the United States (CFIUS) export compliance regulations. Accordingly, IP-based geoblocking is in place for the following countries/territories:

- Cuba
- Iran
- North Korea 
- Syria
- The Crimea, Donetsk People’s Republic, and Luhansk People’s Republic regions of Ukraine

To explore the software for the nodes that currently form the XMTP network, see the [xmtp-node-go repo](https://github.com/xmtp/xmtp-node-go).

Ephemera is in the process of building the replication and incentive mechanisms required to decentralize the network. These mechanisms include:

> 

Author: @neekolas might you be able to 👀 this section below and see if we've captured the work accurately, please and thank you!

- An on-chain smart contract node registry, which lists the currently active set of XMTP nodes
- An off-chain broadcast network of nodes run by a set of independent third parties, which handles the majority of the messaging data (loosely ordered, low latency)

> - A series of on-chain smart contracts on an XMTP L3, which handles account information and the most critical messaging data (strictly ordered, medium latency)

Reviewer 2: Jesse from Base is pretty dogmatic about "onchain" being one word. I don't have a strong opinion, but I think we should do it his way if no one else feels strongly otherwise.

To explore the software for the nodes that currently form the XMTP network, see the [xmtp-node-go repo](https://github.com/xmtp/xmtp-node-go).

> Ephemera is working toward a phased decentralization of the network. Have questions or feedback regarding decentralization? Post to the [XMTP Community Forums](https://community.xmtp.org/c/development/ideas/54).

Reviewer 3: "Working towards a phased approach..." sounds very "not now" to me. I wonder if we should stop referring to the work as phased and just state what we are working on (replication) and discussing (incentives). Wdyt @neekolas @richardhuaaa?

Reviewer: That makes sense to me. Maybe "Ephemera is currently in the process of building the replication and incentive mechanisms for a decentralized network." Depending on how we feel we could link to the repo/tracking task, at least until the XIP/blurb is ready

Author: Thank you for calling this out, @saulmc!

And I love your proposed language, @richardhuaaa! Is this the tracking task you mentioned? https://github.com/xmtp/xmtpd/issues/118

Reviewer: Sample blurb:
This will be comprised of the following components:
- An on-chain smart contract node registry, which lists the currently active set of XMTP nodes
- An off-chain broadcast network of nodes run by independent third parties, for a majority of the messaging data (loosely ordered, low latency)
- A series of on-chain smart contracts on an XMTP L2, for account information and the most critical messaging data (strictly ordered, medium latency)

Author: Thank you both for your feedback here, @saulmc and @richardhuaaa! I've updated the content here: https://github.com/xmtp/docs-xmtp-org/pull/26/files#diff-0f5558e2ccc886a61098eac3e4020166d6c4dee87249fc7dcda8eba0f5bd1196R358

To explore the software for the nodes that currently form the XMTP network, see the [xmtp-node-go repo](https://github.com/xmtp/xmtp-node-go).

Ephemera is in the process of building the replication and incentive mechanisms required to decentralize the network. These mechanism include:

- An on-chain smart contract node registry, which lists the currently active set of XMTP nodes
- An off-chain broadcast network of nodes run by a set of independent third parties, which handles the majority of the messaging data (loosely ordered, low latency)

> - A series of on-chain smart contracts on an XMTP L2, which handles account information and the most critical messaging data (strictly ordered, medium latency)

Reviewer: ```suggestion
- A series of on-chain smart contracts on an XMTP L3, which handles account information and the most critical messaging data (strictly ordered, medium latency)
```
My bad!

> To follow and provide feedback on this work, see the [Replication tracking task](https://github.com/xmtp/xmtpd/issues/118) in the `xmtpd` repo.

Reviewer: LGTM, looping in @neekolas to sign off on this section!

### Will I be able to run my own XMTP node?

Yes, you will be able to run a node. 

Ephemera is in the process of building the replication mechanisms required to enable third-parties to run XMTP nodes and participate in network decentralization.

> To follow and provide feedback on this work, see the [Replication tracking task](https://github.com/xmtp/xmtpd/issues/118) in the `xmtpd` repo.

Reviewer: This is a bit murky depending on the definition of 'node' - the first version of decentralization involves a permissioned set of high-quality originator nodes from third-party organizations decided by a committee. So it's not the case that it's as easy as running a node out of your basement, there are a few additional hoops to jump through. It may be the case though that running a 'gateway node' is frictionless, but we haven't ironed out all of the details there.

I wonder if we should remove the definitive statement `Yes, you will be able to run a node`, which is technically correct but could give the wrong impression.

Author: Thank you for this, @richardhuaaa - again, what thoughtful feedback. I have removed the definitive statement as suggested. And yes, let's add a way for potential third-party organizations and individuals to express interest! Let's add it once we have a clearer picture of how that will work. <3

XMTP provides perceptibly real-time message delivery and retrieval. The network does not provide service level guarantees.

## Fees

> XMTP core developers and researchers are working on outlining a specific fee model for XMTP. The team will share a proposed plan as soon as it’s well-defined enough to generate meaningful input.

Reviewer 3: I like the idea in the other doc about this being more an invitation to discussion than a promise of something later.

Author: Thank you for this feedback, @saulmc - sounds great! I will make the forum topic public and update this section to just point to it.

Reviewer: @neekolas has what seems like a workable approach mapped out, I wonder if we could put two sentences here + invitation to discuss. Example:

XMTP core developers and researchers are working on a specific fee model for XMTP, with the following guiding principles in mind:
- Infrastructure costs for the network must remain low even when decentralized, and comparable to the costs for an equivalent centralized messaging service.
- There must be a sizable free tier for most end users, with the low cost per-user subsidized by the largest apps as well as bulk senders.
- There must be a low 'take rate': the biggest driver of cost must be infrastructure costs.

Author: Thank you for this feedback as well, @saulmc and @richardhuaaa! I feel like I hit the doc review jackpot today! I've updated the content as follows: https://github.com/xmtp/docs-xmtp-org/pull/26/files#diff-0f5558e2ccc886a61098eac3e4020166d6c4dee87249fc7dcda8eba0f5bd1196R414

Reviewer 2: I like that suggestion, but I wouldn't set unreasonable expectations around a "free tier". That's really up to the applications and ecosystem to determine. The protocol expects $$$.

Maybe we just strike that part and replace with: "Fees will be dynamic to protect the network from Denial Of Service attacks"

Reviewer 3: Agreed. Free tier is not part of the protocol design.

To learn more about these keys, see [Key generation and usage in XMTP](/protocol/v2/key-generation-and-usage).

With XMTP v3, needing to sign is no longer based on new sessions, but rather on new app instances.

For example, a first-time user of XMTP first signs with a wallet signature to claim an inbox ID. They also provide a wallet signature to certify that the app-generated installation key is allowed to act on behalf of the inbox ID and can therefore be added to the inbox log's set of authorized keys. These signatures must be provided for each new app instance, or installation.

To learn more about the signatures required for XMTP v3, see [How does XMTP establish a secure and fraud-proof relationship between identities?](#how-does-xmtp-establish-a-secure-and-fraud-proof-relationship-between-identities).

## Storage

### Where are XMTP messages stored?

XMTP stores messages in the XMTP network before and after retrieval. App-specific message storage policies may vary.

### What are the XMTP message retention policies?

#### For XMTP’s blockchain and node databases:

Currently, encrypted payloads are stored indefinitely.

> In the coming year, a retention policy is likely to be added. 

Reviewer 2: I think it's safe to say "will be added"

This retention policy would represent a minimum retention period, not a maximum.

For example, a retention policy may look something like the following, though specifics are subject to change:
  - One year for messages
  - Indefinite storage for account information and personal preferences  

The team is researching a way to provide this indefinite storage and have it scale forever. 
- If research shows that it's possible, we'll share a plan for how it will be achieved. 
- If research shows that it isn't possible, we'll share a plan that shows how retention periods will provide a permanent solution to storage scaling.

Have questions or feedback regarding message storage and retention? Post to the [XMTP Community Forums](https://community.xmtp.org/c/development/ideas/54).

#### For the on-device database managed by the XMTP SDK:

Messages are stored for as long as the user decides to keep them. However, encryption keys are regularly rotated.

### What are the costs of XMTP message storage and retrieval?

Messages are stored off-chain on the XMTP network, with all nodes currently hosted by Ephemera. Ephemera currently absorbs all message storage and retrieval costs.

> There are no message storage and retrieval-related fees incurred by developers for building with the XMTP SDK.

Reviewer 2: Maybe add a "today". We do expect message storage fees to be arriving next year.

XMTP provides perceptibly real-time message delivery and retrieval. The network does not provide service level guarantees.

## Fees

XMTP core developers and researchers are working on a specific fee model for XMTP, with the following guiding principles in mind:

- Infrastructure costs for the network must remain low even when decentralized, and comparable to the costs for an equivalent centralized messaging service.
- There must be a sizable free tier for most end users, with the low cost per-user subsidized by the largest apps and bulk senders.
- There must be a low "take rate": the biggest driver of cost must be infrastructure costs, with any remaining cost returned to the network.

> Have questions or feedback about the fee model for XMTP? See [XMTP fees: Guiding principles and FAQ](https://community.xmtp.org/t/xmtp-fees-guiding-principles-and-faq/795) in the XMTP Community Forums.

Reviewer: Looping in @neekolas on this section to get the messaging right, he's much deeper in this than I am

## Security

### Has XMTP undergone a security audit?

A security audit of [LibXMTP](https://github.com/xmtp/libxmtp) and its use of Messaging Layer Security (MLS) will begin in Q4 2024. The audit will be performed by [NCC Group](https://www.nccgroup.com/).

The [XMTP JavaScript SDK](https://github.com/xmtp/xmtp-js) (`xmtp-js`) was verified by [CertiK](https://www.certik.com/company/about) on Feb 6, 2023. Read their [XMTP security assessment](https://raw.githubusercontent.com/xmtp/docs-xmtp-org/main/docs/pages/img/REP-final-20230207T000355Z.pdf).

> ### How does XMTP establish a secure and fraud-proof relationship between identities?

Reviewer: LGTM, the only comment would be to make sure that app developers know that this is mostly handled by the SDK under the hood

With XMTP, the concept of an identity centers on an inbox ID that has a constantly changing set of identities that can be either wallets or app installations.

1. A user starts with an Ethereum wallet and installs an XMTP-enabled app.

> 2. The user creates an XMTP inbox ID, which can represent a set of identities (wallets and appinstallations).

Reviewer: ```suggestion
2. The user creates an XMTP inbox ID, which can represent a set of identities (wallets and app installations).
```

3. The first wallet to claim an inbox ID becomes the first member of the set and is designated as the "recovery address."

4. For each app installation, an installation key is generated by the app.

5. Additional wallets and app installations can be added to the set through a two-way signature process:
   - A signature from a key currently in the set
   - A signature from the key being added to the set

6. Wallets and app installations can be removed from the set using a signature from the recovery address.

> 7. Changes to an inbox ID's set of identities are published as "identity updates" to the server, which stores them as an "inbox log."

Reviewer: ```suggestion
7. Changes to an inbox ID's set of identities are published as "identity updates" to the XMTP network, which stores them as an "inbox log."
```

MLS credentials contain only the inbox ID. When a client receives a credential, they verify it by:
- Downloading the inbox log for that inbox ID
- Reading through the series of updates
- Verifying that the installation key is a current valid member of the set

For group chats, XMTP uses the MLS protocol to manage secure communication between multiple parties.

For a first-time XMTP user, the process typically involves:
- One wallet signature to claim the Inbox ID and add the installation key (these can be batched)
- One installation key signature (handled invisibly by the app) to consent to being added to the inbox ID.

> This process requires only one visible signature from the end user, in addition to any XMPT v2 signatures.

Reviewer: ```suggestion
This process requires only one visible signature from the end user, in addition to any XMTP v2 signatures.
```

With XMTP v3, needing to sign is no longer based on new sessions, but rather on new app instances.

For example, a first-time user of XMTP first signs with a wallet signature to claim an inbox ID. They also provide a wallet signature to certify that the app-generated installation key is allowed to act on behalf of the inbox ID and can therefore be added to the inbox log's set of authorized keys. These signatures must be provided for each new app instance, or installation.

To learn more about the signatures required for XMTP v3, see [How does XMTP establish a secure and fraud-proof relationship between identities?](#how-does-xmtp-establish-a-secure-and-fraud-proof-relationship-between-identities).

## Storage

### Where are XMTP messages stored?

XMTP stores messages in the XMTP network before and after retrieval. App-specific message storage policies may vary.

### What are the XMTP message retention policies?

#### For XMTP’s blockchain and node databases:

Currently, encrypted payloads are stored indefinitely.

In the coming year, a retention policy is likely to be added. 

> This retention policy would represent a minimum retention period, not a maximum. For example, the policy will provide a functional guarantee not a privacy guarantee. A privacy guarantee is more difficult to provide due to the decentralized nature of the network and subtleties of blockchain data availability.

Reviewer: ```suggestion
This retention policy would represent a minimum retention period, not a maximum.
```

### What are the costs of XMTP message storage and retrieval?

Messages are stored off-chain on the XMTP network, with all nodes currently hosted by Ephemera. Ephemera currently absorbs all message storage and retrieval costs.

There are no message storage and retrieval-related fees incurred by developers for building with the XMTP SDK.

## Messages

### Which message formats does XMTP support?

XMTP transports a message payload as a set of bytes that can represent any format that a developer wants to support, such as plain text, JSON, or non-text binary or media content.

With XMTP, these message formats are called content types. Currently, there are two basic content types available. These content types aim to establish broad compatibility among apps built with XMTP.

The XMTP community can propose and adopt standards for other content types, either informally or through a governance process.

To learn more about content types, see [Content types](/content-types/content-types).

To learn more about the XMTP improvement proposals governance process, see [What is an XIP?](https://github.com/xmtp/XIPs/blob/main/XIPs/xip-0-purpose-process.md)

> ### Which message metadata does XMTP support?

Author: Correct in v2 and v3?

Reviewer: I'm unsure about the intent of this question - is it intended to be about what extra metadata apps are able to attach to message payloads, or is it about what metadata gets leaked onto the network?

Author: Thank you for this question, @richardhuaaa - it is the latter - are there any changes to the types of metadata leaked onto the network? =)

Reviewer: This one is a bit complex and requires a lot of precision - I wonder if it would be best to link to a section in our [audit docs](https://hackmd.io/@jan-cryspen/rJ5W-83j0#Privacy) for now, @neekolas. The [MLS spec](https://www.rfc-editor.org/rfc/rfc9420.html#section-16.4) also has some general info on MLS but it may not be enough for our specific implementation.

There are no messaging-related fees incurred by developers for building with the XMTP SDK.

## Security

### Has XMTP undergone a security audit?

A security audit of [LibXMTP](https://github.com/xmtp/libxmtp) and its use of Messaging Layer Security (MLS) will begin in Q4 2024. The audit will be performed by [NCC Group](https://www.nccgroup.com/).

The [XMTP JavaScript SDK](https://github.com/xmtp/xmtp-js) (`xmtp-js`) was verified by [CertiK](https://www.certik.com/company/about) on Feb 6, 2023. Read their [XMTP security assessment](https://raw.githubusercontent.com/xmtp/docs-xmtp-org/main/docs/pages/img/REP-final-20230207T000355Z.pdf).

### How does XMTP establish a secure and fraud-proof relationship between two identities?

> XMTP uses a combination of Ethereum wallet signatures and MLS (Messaging Layer Security):

Author: is this an accurate summary of how v3 works?

Reviewer: So there is no identity key anymore - there are only wallets and installation keys. Also, the format of the credential has changed significantly after the multi-identity work, happy to run through it verbally!

A rough intuition:
- Each Inbox ID represents a constantly changing 'set' of identities, that could be either wallets or installations.
- The first wallet to claim an Inbox ID is the first member of the set, and is also the 'recovery address'.
- More wallets and installations can be added to the set via two-way signatures (a signature from a key currently in the set, accompanied by a signature from the key being added to the set)
- Wallets and installations can be removed from the set using a signature from the 'recovery address' (one-way signature)
- These changes to the set are published as 'identity updates' to the server, which stores the complete set of updates for any inbox ID as an 'Inbox log'.

Credentials no longer contain anything other than an Inbox ID. Whenever any other client receives your credential, they will verify it by downloading the 'Inbox log' for that Inbox ID, reading through the series of updates, and verifying at the end that your installation key is a currently valid member of the 'set'.

As an example, if you are a first time user of XMTP, these will be the signatures required:
1. Wallet signature to claim the Inbox ID
2. Wallet signature to add the installation key to the inbox (which is generated by the app)
3. Installation key signature to consent to being added

We have 'batch signatures', so a single end-user wallet signature can sign for both of the first two points. For the third signature, this happens inside the app invisibly to the user (intuitively: wallet signatures represent user actions, installation signatures represent app actions).

Author: What a generous comment from you, @richardhuaaa - thank you!

I have made these updates to reflect the info you provided: https://github.com/xmtp/docs-xmtp-org/pull/26/files#diff-0f5558e2ccc886a61098eac3e4020166d6c4dee87249fc7dcda8eba0f5bd1196R432

WDYT?

Beyond this first pass, I can do more to bring this FAQ in line with [Scott's excellent work here](https://docs.xmtp.org/protocol/v3/identity), which I know you have already invested time in.

Reviewer: LGTM!

4. The installation key is used to create an MLS credential, which is signed by the Ethereum wallet.

5. The MLS credential contains the wallet address and a hash of the installation's signature verification key.

6. For group chats, XMTP uses the MLS protocol to manage secure communication between multiple parties.

For more details, see [Group chat concepts and protocols](/protocol/v3/group-chat#security-and-encryption) and [Identity in XMTP v3](/protocol/v3/identity)

#### XMTP V2

In V2, blockchain accounts sign and advertise a set of keys to start using XMTP. These keys are used to establish a shared secret between the blockchain accounts. The shared secret is then used to generate a key for encrypting an invitation, allowing the blockchain accounts to start exchanging messages. No third-party apps or relayers are involved in this process.

For more details, see [Key generation and usage in XMTP v2](/protocol/v2/key-generation-and-usage) and [Invitation and message encryption with XMTP v2](/protocol/v2/invitation-and-message-encryption).

### Does each blockchain account address have a corresponding XMTP identity?

> Yes, each blockchain account address is associated with an XMTP identity.

Author: is this an accurate summary of how v3 works?

Author: Based on @richardhuaaa's review, I updated this section as follows: https://github.com/xmtp/docs-xmtp-org/pull/26/files#diff-0f5558e2ccc886a61098eac3e4020166d6c4dee87249fc7dcda8eba0f5bd1196R473

1. When a user first uses XMTP with a particular blockchain account, they create an "identity key" for that account.

> 2. This identity key is used to sign "installation keys" for each app or device where the user uses XMTP.

Reviewer 3: Multi-identity could be explained pretty succinctly this way too. Something like

"The identity key is also used to add or remove account access for additional wallets."

Author: Great - thank you for this important callout, @saulmc! I believe the guidance Rich provided has enabled the FAQ to more directly address multi-wallet identity.  Let me know if you have any concerns! I'm all ears, as always.....

3. The installation keys are then used to create and sign MLS credentials, which are used for secure communication in individual and group chats.

To learn more about identity in XMTP v3, see [Identity in XMTP v3](/protocol/v3/identity).

#### XMTP V2

Yes. Each blockchain account address is represented by an XMTP identity key. This identity key is a part of a key bundle that only the address that generated it can use to authenticate messages.

To learn more about XMTP identity keys, see [Key generation and usage in XMTP](/protocol/v2/key-generation-and-usage).

### Why do apps built with XMTP require a user to sign with their blockchain account private keys each time they start a new messaging session?

> Users need to authenticate with their blockchain account when starting a new messaging session in this way:

Author: is this an accurate summary of how v3 works?

Reviewer: Similar to above - it's close, but we've nixed the identity key.

First session: Wallet signature to claim inbox ID

Every session: Wallet signature to certify that the app-generated installation key is allowed to act on behalf of the inbox. (Or put another way, to add the installation key to the inbox log's set of authorized keys)

Author: Thank you for this guidance as well, @richardhuaaa! 

I've made the following updates based on the information you've shared: https://github.com/xmtp/docs-xmtp-org/pull/26/files#diff-0f5558e2ccc886a61098eac3e4020166d6c4dee87249fc7dcda8eba0f5bd1196R483

WDYT? 

No rush on any of these for today, of course! Perhaps you might be able to take a look on Monday? Thank you again!

Reviewer: LGTM!

## Messages

### Which message formats does XMTP support?

XMTP transports a message payload as a set of bytes that can represent any format that a developer wants to support, such as plain text, JSON, or non-text binary or media content.

With XMTP, these message formats are called content types. Currently, there are two basic content types available. These content types aim to establish broad compatibility among apps built with XMTP.

The XMTP community can propose and adopt standards for other content types, either informally or through a governance process.

To learn more about content types, see [Content types](/content-types/content-types).

To learn more about the XMTP improvement proposals governance process, see [What is an XIP?](https://github.com/xmtp/XIPs/blob/main/XIPs/xip-0-purpose-process.md)

> ### Which message metadata does XMTP support?

Author: Here's where Cursor hallucinated about how XMTP handles metadata in v3 vs v2. I thought it was interesting - so sharing in this PR.

Would love to understand if there are differences in metadata handling in v2 vs v3.
